locations/location_mgr.go

```diff
@@ -0,0 +1,379 @@
+// Copyright 2019 Bull S.A.S. Atos Technologies - Bull, Rue Jean Jaures, B.P.68, 78340, Les Clayes-sous-Bois, France.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package locations is responsible for handling locations where deployments
+// can take place
+package locations
+
+import (
+	"encoding/json"
+	"github.com/ystia/yorc/v4/locations/adapter"
+	"github.com/ystia/yorc/v4/prov/hostspool"
+	"path"
+	"time"
+
+	"github.com/hashicorp/consul/api"
+	"github.com/pkg/errors"
+	"github.com/spf13/viper"
+
+	"github.com/ystia/yorc/v4/config"
+	"github.com/ystia/yorc/v4/deployments"
+	"github.com/ystia/yorc/v4/helper/consulutil"
+	"github.com/ystia/yorc/v4/log"
+	"github.com/ystia/yorc/v4/tosca"
+)
+
+// LocationConfiguration holds a location configuration
+type LocationConfiguration struct {
+	Name       string            `yaml:"name" mapstructure:"name"`
+	Type       string            `yaml:"type,omitempty" mapstructure:"type"` // not an enum as it could be extended by plugins
+	Properties config.DynamicMap `yaml:"properties,omitempty" mapstructure:"properties"`
+}
+
+// A Manager is in charge of creating/updating/deleting locations from the pool
+type Manager interface {
+	InitializeLocations(locationFilePath string) (bool, error)
+	CreateLocation(lConfig LocationConfiguration) error
+	RemoveLocation(locationName, locationType string) error
+	SetLocationConfiguration(lConfig LocationConfiguration) error
+	GetLocations() ([]LocationConfiguration, error)
+	GetLocationProperties(locationName, locationType string) (config.DynamicMap, error)
+	GetLocationPropertiesForNode(deploymentID, nodeName, locationType string) (config.DynamicMap, error)
+	GetPropertiesForFirstLocationOfType(locationType string) (config.DynamicMap, error)
+	Cleanup() error
+}
+
+type locationManager struct {
+	cc        *api.Client
+	hpAdapter adapter.LocationAdapter
+}
+
+// LocationsDefinition represents the structure of an initialization file defining locations
+type LocationsDefinition struct {
+	Locations []LocationConfiguration `yaml:"locations" mapstructure:"locations" json:"locations"`
+}
+
+// NewManager creates a Location Manager
+func NewManager(cfg config.Configuration) (Manager, error) {
```

Comment:
It is was not clear for me at the first glance that the manager may be cached.
Maybe if the function was named `GetManager` I may have a doubt and checked it.